toolchain/check/handle_impl.cpp

```diff
@@ -206,22 +259,48 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
 
   // TODO: Check for an orphan `impl`.
 
-  // TODO: Check parameters. Store them on the `Impl` in some form.
-  static_cast<void>(params_id);
+  // Add the impl declaration.
+  SemIR::ImplDecl impl_decl = {.impl_id = SemIR::ImplId::Invalid,
+                               .decl_block_id = decl_block_id};
+  auto impl_decl_id =
+      context.AddPlaceholderInst(SemIR::LocIdAndInst(node_id, impl_decl));
+
+  SemIR::Impl impl_info = {
+      name_context.MakeEntityWithParamsBase(name, impl_decl_id,
+                                            /* is_extern=*/false,
+                                            SemIR::LibraryNameId::Invalid),
+      {.self_id = self_type_id, .constraint_id = constraint_type_id}};
 
   // Add the impl declaration.
   // TODO: Does lookup in an impl file need to look for a prior impl declaration
   // in the api file?
-  auto impl_id = context.impls().LookupOrAdd(self_type_id, constraint_type_id);
-  SemIR::ImplDecl impl_decl = {.impl_id = impl_id,
-                               .decl_block_id = decl_block_id};
-  auto impl_decl_id = context.AddInst(node_id, impl_decl);
+  auto& impl_lookup =
+      context.impls().LookupBucket(impl_info.self_id, impl_info.constraint_id);
+  for (auto prev_impl_id : impl_lookup) {
+    if (MergeImplRedecl(context, impl_info, prev_impl_id)) {
+      impl_decl.impl_id = prev_impl_id;
+      break;
+    }
+  }
+
+  bool is_new_impl = !impl_decl.impl_id.is_valid();
+  if (is_new_impl) {
```

Comment:
```suggestion
  // Create a new impl if this isn't a valid redeclaration.
  if (!impl_decl.impl_id.is_valid()) {
```

Comparing with similar logic in other entities (class/function/interface), with `is_new_class`, the variable is reused. Here, this is the only use of `is_new_impl`; the declaration made me look, which is why I'm suggesting eliding it. But then I expect the variable name is somewhat documenting, so suggesting the comment similar to what we have elsewhere.